Cherify dlmalloc and add revocation support

brooksdavis opened the pull request:
Return appropriately bounded capabilities for each allocation.

brooksdavis committed:
Disable the shadow map stuff in pure-cap mode.

It relies on contructing pointers from thin air.  We need to switch
to a model of hanging shadow maps off of segments (and some segment
equivalent for mmaped allocations).

brooksdavis committed:
Use alignment builtins.

brooksdavis committed:
Prevent direct mmap allocation of large data.

We to be able to get back to the mapped region from a pointer in the
free() path.  Forcing everything to be a segment is the easy solution.

brooksdavis committed:
Disable merging of segments in purecap mode.

We can't merge capabilities so we need each segment to stand alone even
if it's adjacent to another one.

brooksdavis committed:
Remove the bounds from pointers passed to free() and realloc().

brooksdavis committed:
Make sure FOOTERS isn't turned on.

When it's enabled, there's a circular dependency between unbounding
pointers and finding the mstate pointer.

brooksdavis committed:
Make a real internal_malloc().

When start bounding pointers we don't want internal_malloc() to return
bounded pointers because other callers (internal_memalign()) need to
tinker with the underlying chunk structures and it's expensive to
unbound pointers.

brooksdavis committed:
Put bounds on pointers returned to the caller.

Round up all request to a representable length at the top level call.
The internal functions can then assume the requests are representable.

brooksdavis committed:
Pad struct malloc_chunk to a power of two.

brooksdavis committed:
Remove an untrue (and seemingly unnecessicary) assertion.

I've not found any cases where size_t and pointers are interchanged
(other then the casts through size_t removed in a previous commit).

brooksdavis committed:
Tidy the makefile and assume that LD is clang.

Add a dumb test program that allocates some memory and prints the
pointers.

brooksdavis committed:
Add a pad element to malloc_tree_chunk matching malloc_chunk.

brooksdavis committed:
Move the pad outside the malloc region.

brooksdavis committed:
Store a copy of the capability pointer we return in the pad.

This provides stronger assurance that we're actually freeing a chunk.
I will also use the for revocation rather than trusting that the
capability passed to free() has the right bounds or rebuilding a one
from the chunk metadata.

A somewhat more advanced configuration would seal the capability
providing strong assurance that a fake malloc_chunk has not be
maliciously constructed.

While here, use USAGE_ERROR_ACTION on all lookup failures in
unbound_ptr().

brooksdavis committed:
Require that the pointer be tagged.

brooksdavis committed:
Add revocation support.

This extracts the existing set and clear logic into a new function,
dlmalloc_revoke().  Shadow buffer pointers are hung off each segment.
To avoid looking up the segment the check is in up to three times for
each revoked chunk, store a pointer to the shadow buffer in the pad
element of the chunk.  A single lookup is required in all cases to
allow us to remove bounds from the pointer in order to access the chunk.

brooksdavis committed:
Abstract out the chunk header offset.

brooksdavis committed:
Update unbound_ptr to optionaly return the segment.

brooksdavis committed:
Add an option to move the whole struct malloc_chunk outside the
allocation.

We need a doubly linked list for efficient consolidation of segments in
quarintine and it's convenient to be able to hold a pointer to the
segment.

brooksdavis committed:
Avoid some unnecessicary UB (taking the address of something derived
from NULL).

brooksdavis committed:
Work around a possible initilization issue in freebin.

brooksdavis committed:
Actually save enough space for a chunk plus two fenceposts.

This commit is enough to let use install the msegement object, but
something isn't quite right.  Performing large allocations (1MB) causes
corruption of the sp pointer in an msegment and the segment_holding
crashes when called from unbound_ptr in free().

brooksdavis committed:
Handle allocations requiring stronger than default alignment.

When such allocations are requested, use internal_memalign() instead of
internal_malloc() to perform allocation.

brooksdavis committed:
Add some (overly local) CC and LD settings.

Can't use ?= for the assignments due to system defaults, but you can
still override on the command line.

brooksdavis committed:
Make it more clear that an insert_freebuf_chunk() always occurs in the
consolidation code.

This is either the new chunk (possiably extended with the next chunk) or
the previous chunk which was removed after merging the new chunk into it.

brooksdavis committed:
Clear the pad pointer in dlfree.

This will catch double-free attempts in unbound_ptr.

brooksdavis committed:
Start fenceposts in the right place.

We've allocated ssize space for the chunk. We need to increment all the
way past it, not ssize-chunk overhead past it.  This fixes trashing the
shadow buffer pointer in new segments.

brooksdavis committed:
Correct a typo reported by @arichardson.

brooksdavis committed:
Lock around manual calls to malloc_revoke.

brooksdavis committed:
Actually allocate enough space for the request.

Previously a combination of accounting for too little overhead and an
improper computation of MIN_CHUNK_SIZE (negative) resulated in
pad_request() computing chunk sizes that were too small.  This resulted
in allocation bounds overlapping chunk headers.

The new compuation of MIN_CHUNK_SIZE is slightly larger than it needs to
be (a full 64 bytes of payload), but we're always paying at least a
64-byte cost so it probably doesn't make sense to allow very slightly
smaller payloads.

brooksdavis committed:
Assert that we've allocated a large enough chunk.

brooksdavis committed:
Zero all allocations before returning them.

This ensures that the transitive-closure of pages reachable from a
given point in the code only increases by the allocation. The actual
implementation is quite inefficent and performs unnecessicary zeroing
for newly mapped memory, values used by realloc, and (eventually) pages
that are entierly revoked and are now zeroed as a result.

nwf commented on dlmalloc_nonreuse.c:
```diff
@@ -1195,6 +1230,59 @@ static int has_segment_link(mstate m, msegmentptr ss) {
+	/*
+	 * Make sure we're inbounds.  Otherwise our address might match
+	 * the wrong segment.
+	 */
```
You can elide the bounds test, but not the taggedness test, if you use the base, since that can't ever go out of bounds of the original allocation.

brooksdavis replied:
I ended up deciding to check that offset == 0 since anything else is a bug anyway.  Using base in segment_holding also probably makes sense, but this lets us trivially detect another type of bug.

nwf commented on dlmalloc_nonreuse.c:
```diff
@@ -3219,52 +3394,71 @@ dlfree(void* mem) {
+  }
+
+  struct caprevoke_stats crst;
+  caprevoke(CAPREVOKE_LAST_PASS|CAPREVOKE_IGNORE_START, 0, &crst);
```
This should probably, strictly, be something like the following, just in case someone else is doing revocation in the same process on a different thread as well.  Sorry the wiki doc didn't actually spell this out; got lost in all the rest of the writing.

```
struct caprevoke_stats crst;
uint64_t start_epoch;
caprevoke(CAPREVOKE_LAST_PASS|CAPREVOKE_IGNORE_START|CAPREVOKE_NO_WAIT_OK, 0, &crst);
start_epoch = crst.epoch_init;
while (!caprevoke_epoch_clears(crst->epoch_fini, start_epoch)) {
  caprevoke(CAPREVOKE_LAST_PASS, start_epoch, &crst);
}
```

It's very likely that the `caprevoke_epoch_clears` test reveals that no more calls are necessary; perhaps a `__builtin_expect` is useful.

nwf replied:
Hm.  In some cases that I don't understand, the looping form is actually necessary; we get some bad return codes from the vm in `caprevoke` that cause us to bail back to userland but if we just retry they eventually clear up.  That's not great, but it's just once at the process startup, it seems.

nwf reviewed (commented):
Oh this is so very very exciting!

Just some nits; I'm sorry about the `caprevoke` API being mostly in my head.

brooksdavis committed:
Revert "Actually save enough space for a chunk plus two fenceposts."

Fixing pad_request() fixed this issue.  The magic numbers here remain
concerning...

This reverts commit a0b53ee37352df3e664f3975fa9bc98b5b758910.

brooksdavis committed:
Make sure that we round large lengths up sufficently.

This is particularly important for underlying mmap calls where 64K could
potentially not be sufficent alignment.

brooksdavis committed:
Rather than checking the pointer is inbounds, check that offset == 0.

It is always a bug for malloc to be called with a non-zero offset.

brooksdavis committed:
Use a caprevoke() loop to allow concurent revokers.

Submitted by: @nwf

brooksdavis committed:
Revert "Revert "Actually save enough space for a chunk plus two fenceposts.""

This reverts commit adb00c4f251139b7943a4b3318b71505f82152ae.

brooksdavis committed:
Use #if 0 to disable the shrinking code.

I kept missing where the termination was and getting confused...

brooksdavis committed:
Return early in easy cases to avoid bounding NULL or already-bound
pointers.

brooksdavis committed:
Attempt to handling capability compression in realloc.

Only try extending in place if the current alignment is sufficient to
support the new size.

When using a new allocation, make sure to request that it be aligned
appropriatly if required.

brooksdavis committed:
Make the tests actually check things instead of printing pointers for
manual checking.
